### PR TITLE
implement ListOps.all and ListOps.any

### PR DESCRIPTION
**Pull request by odeac**
## Overview
See title. They are needed for some PoS tests

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3463

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/test/resources/ListOpsTest.rho [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    contract test_all(rhoSpec, _, ackCh) = {
+      new ch0, ch1, ch2, ch3, ch4, greaterThan1 in {
+        contract greaterThan1(@x, retCh) = { retCh!(x > 1)} |
+
+        @ListOps!("all", [], *greaterThan1, *ch0) |
+        @ListOps!("all", [1], *greaterThan1, *ch1) |
+        @ListOps!("all", [2], *greaterThan1, *ch2) |
+        @ListOps!("all", [3, 4], *greaterThan1, *ch3) |
+        @ListOps!("all", [1,2, 3, 4], *greaterThan1, *ch4) |
```
Comment (KentShikama) [inline casper/src/test/resources/ListOpsTest.rho]: nitpick: 1,2 formatting

File: casper/src/test/resources/ListOpsTest.rho [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            ((true, "== <-", *ch0), "all is successful for empty lists"),
+            ((false, "== <-", *ch1), ""),
+            ((true, "== <-", *ch2), ""),
+            ((true, "== <-", *ch3), ""),
+            ((false, "== <-", *ch4), "")
+          ], *ackCh)
+      }
+    } |
+
+    contract test_all_is_lazy(rhoSpec, _, ackCh) = {
```
Comment (KentShikama) [inline casper/src/test/resources/ListOpsTest.rho]: This is short_circuit not lazy

File: casper/src/test/resources/ListOpsTest.rho [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    } |
+
+    contract test_any(rhoSpec, _, ackCh) = {
+      new ch0, ch1, ch2, ch3, ch4 in {
+
+        @ListOps!("any", [], *greaterThan1, *ch0) |
+        @ListOps!("any", [1], *greaterThan1, *ch1) |
+        @ListOps!("any", [2], *greaterThan1, *ch2) |
+        @ListOps!("any", [3, 4], *greaterThan1, *ch3) |
+        @ListOps!("any", [1,2, 3, 4], *greaterThan1, *ch4) |
```
Comment (KentShikama) [inline casper/src/test/resources/ListOpsTest.rho]: same nitpick as above

File: casper/src/test/resources/ListOpsTest.rho [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            ((false, "== <-", *ch0), "any returns false for empty lists"),
+            ((false, "== <-", *ch1), "[1]"),
+            ((true, "== <-", *ch2), "[2]"),
+            ((true, "== <-", *ch3), "[3,4]"),
+            ((true, "== <-", *ch4), "[1,2,3,4]")
+          ], *ackCh)
+      }
+    } |
+
+    contract test_any_is_lazy(rhoSpec, _, ackCh) = {
```
Comment (KentShikama) [inline casper/src/test/resources/ListOpsTest.rho]: same as above

File: casper/src/test/resources/ListOpsTest.rho (hunk 2 above)
Comment (odeac) [reply casper/src/test/resources/ListOpsTest.rho]: Of course! I couldn't remember how it's called :)

File: casper/src/main/resources/ListOps.rho [hunk 5]
```diff
@@ -175,6 +175,20 @@ new  ListOps, rs(`rho:registry:insertSigned:secp256k1`), uriOut in {
     }
   } |
 
+  contract ListOps(@"all", @list, predicate, returnCh) = {
```
Comment (JosDenmark) [inline casper/src/main/resources/ListOps.rho]: This process should have a more informative name. Something like `forall` is consistent with Scala and with mathematical vocabulary generally.

File: casper/src/main/resources/ListOps.rho [hunk 6]
```diff
@@ -193,6 +193,25 @@ new  ListOps, rs(`rho:registry:insertSigned:secp256k1`), uriOut in {
...
   } |
 
+  contract ListOps(@"any", @list, predicate, returnCh) = {
+    new combine, resultCh, stdout(`rho:io:stdout`) in {
+      ListOps!("partialFold", list, false, *combine, *resultCh) |
+      contract combine(@item, @acc, accResultCh) = {
+        new predResultCh in {
+          predicate!(item, *predResultCh) |
+          for (@isSatisfied <- predResultCh) {
+            stdout!({"item": item, "acc": acc, "isSatisfied": isSatisfied}) |
```
Comment (JosDenmark) [inline casper/src/main/resources/ListOps.rho]: Will you Command+F for appearances of `stdout` in the next changes you make to Rholang code?

File: casper/src/main/resources/ListOps.rho [hunk 7]
```diff
@@ -193,6 +193,25 @@ new  ListOps, rs(`rho:registry:insertSigned:secp256k1`), uriOut in {
     }
   } |
 
+  contract ListOps(@"any", @list, predicate, returnCh) = {
```
Comment (JosDenmark) [inline casper/src/main/resources/ListOps.rho]: We don't have to mimic scala method naming conventions but this should have a more informative name. This should have a more informative name. Scala and Haskell use `exists`.

File: casper/src/test/resources/ListOpsTest.rho [hunk 8]
```diff
@@ -61,6 +65,50 @@ in {
...
+          }
+        } |
+
+        @ListOps!("any", [2, 1, 0], *greaterThan1WithCount, *resultCh) |
+
+        for (@result <- resultCh) {
+          rhoSpec!("assertMany",
+            [
+              ((true, "==", result), "first item satisfies the predicate"),
+              ((1, "== <-", *callCountCh), "after the first failure no more calls should happen"),
```
Comment (JosDenmark) [inline casper/src/test/resources/ListOpsTest.rho]: If we're checking whether there exists an element of the collection that satisfies a predicate, why should we short-circuit when one element fails? It should actually be the opposite. The call short-circuits after the first success. Similarly,`forall` should only short-circuit after the first failure.

File: casper/src/main/resources/ListOps.rho (hunk 6 above)
Comment (odeac) [reply casper/src/main/resources/ListOps.rho]: one day...

File: casper/src/test/resources/ListOpsTest.rho (hunk 8 above)
Comment (odeac) [reply casper/src/test/resources/ListOpsTest.rho]: This is a typo in the message. The test is correct

**Pull request merged**